### ISSUE TITLE
ImmutablePOJOMapper cannot map more than one nested POJO on the same level

### ISSUE THREAD

**Issue by strindberg** [jOOQ/jOOQ#12327]
### Expected behavior

We have some model objects which contain both "grandparent"-objects (i.e. objects that are connected through foreign keys in the central object) and "child"-objects, which are composed from columns within the central table. We would expect these objects to be read into model objects.

### Actual behavior
"grandparent"-objects and "child"-objects can be read in isolation, but when both are present in the central object, some sort of mismatch in the jOOQ mapper results in the constructors being called with wrong parameter types. It seems as though the code that finds the correct mapper for each sub object get misaligned, and the wrong objects are given to the constructors.

### Steps to reproduce the problem

Database definitions:

```sql
CREATE TABLE grand_parent (
    id     bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    name   text   NOT NULL,

    PRIMARY KEY (id)
);

CREATE TABLE parent (
    id           bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    gp_id        bigint NOT NULL,
    child_name   text   NOT NULL,
    child_descr  text   NOT NULL,

    PRIMARY KEY (id),
    FOREIGN KEY (gp_id) REFERENCES grand_parent (id)
);
```

Kotlin model:

```kotlin
data class GrandParent(val id: Long, val name: String)

data class Parent(val id: Long, val gp: GrandParent, val child: Child)

data class Child(val name: String, val descr: String)

```

Repository code (Kotlin):
```kotlin
    fun getById(id: Long): Parent? =
        dsl.select(
            PARENT.ID,
            PARENT.grandParent().ID.`as`("gp.id"),
            PARENT.grandParent().NAME.`as`("gp.name"),
            PARENT.CHILD_NAME.`as`("child.name"),
            PARENT.CHILD_DESCR.`as`("child.descr")
        ).from(PARENT)
            .where(PARENT.ID.eq(id))
            .fetchOne()
            ?.into(Parent::class.java)
```

### Versions

- jOOQ: 3.15.1
- Java: 11
- Kotlin: 1.5.21
- Database (include vendor): Postgres 13
- OS: MacOS, Linux
- JDBC Driver (include name if inofficial driver):  org.postgresql:postgresql:42.2.23

**Comment by lukaseder** [comment]
Thanks a lot for your report. I will look into this soon.

**Comment by strindberg** [comment]
A MCVE is now available here: https://github.com/strindberg/jOOQ-mcve.git, branch `jooq_issue_12327`, sub-project `
jOOQ-mcve-kotlin `.

**Comment by lukaseder** [comment]
Thanks for the MCVE. I can reproduce this exception. Is that the one you're seeing?

```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.408 sec <<< FAILURE!
mcveTest(org.jooq.mcve.test.kotlin.KotlinTest)  Time elapsed: 1.351 sec  <<< ERROR!
org.jooq.exception.MappingException: An error ocurred when mapping record to class org.jooq.mcve.test.kotlin.KotlinTest$Parent
        at org.jooq.impl.DefaultRecordMapper$ImmutablePOJOMapper.map(DefaultRecordMapper.java:1127)
        at org.jooq.impl.DefaultRecordMapper.map(DefaultRecordMapper.java:610)
        at org.jooq.impl.AbstractRecord.into(AbstractRecord.java:804)
        at org.jooq.mcve.test.kotlin.KotlinTest.getById(KotlinTest.kt:50)
        at org.jooq.mcve.test.kotlin.KotlinTest.mcveTest(KotlinTest.kt:66)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
Caused by: java.lang.IllegalArgumentException: argument type mismatch
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at org.jooq.impl.DefaultRecordMapper$ImmutablePOJOMapper.map(DefaultRecordMapper.java:1124)
        ... 38 more
```

**Comment by lukaseder** [comment]
It seems that both nested records are mapped into the same `GrandParent` type:

![image](https://user-images.githubusercontent.com/734593/134155601-7a1669ec-8d35-4d7f-9df8-873f95bbbac8.png)

**Comment by lukaseder** [comment]
I can reproduce the problem in Java like this:

```java
    @Test
    public void testNestedImmutablePojosMultipleNests() {
        Field<String> a = field("a", VARCHAR);
        Field<String> b1 = field("inner1.b", VARCHAR);
        Field<String> c1 = field("inner1.inner.c", VARCHAR);
        Field<String> d1 = field("inner1.d", VARCHAR);
        Field<String> e = field("e", VARCHAR);
        Field<String> c2 = field("inner2.c", VARCHAR);

        Record6<String, String, String, String, String, String> rec = create.newRecord(a, b1, c1, d1, e, c2);
        rec.values("a", "b", "c", "d", "e", "f");
        ImmutableOutersMultipleNests pojo = rec.into(ImmutableOutersMultipleNests.class);

        assertEquals("a", pojo.a);
        assertEquals("b", pojo.inner1.b);
        assertEquals("c", pojo.inner1.inner.c);
        assertEquals("d", pojo.inner1.d);
        assertEquals("e", pojo.e);
        assertEquals("f", pojo.inner2.c);
    }

    static class ImmutableOutersMultipleNests {
        final String a;
        final ImmutableInnersMultipleNests1 inner1;
        final String e;
        final ImmutableInnersMultipleNests2 inner2;

        ImmutableOutersMultipleNests(String a, ImmutableInnersMultipleNests1 inner1, String e, ImmutableInnersMultipleNests2 inner2) {
            this.a = a;
            this.inner1 = inner1;
            this.e = e;
            this.inner2 = inner2;
        }
    }

    static class ImmutableInnersMultipleNests1 {
        final String b;
        final ImmutableInner2 inner;
        final String d;

        ImmutableInnersMultipleNests1(String b, ImmutableInner2 inner, String d) {
            this.b = b;
            this.inner = inner;
            this.d = d;
        }
    }

    static class ImmutableInnersMultipleNests2 {
        final String c;

        ImmutableInnersMultipleNests2(String c) {
            this.c = c;
        }
    }
```

**Comment by strindberg** [comment]
Yes, we're seeing the same problem. I should have been more clear on the error, but you have found it: for some reason the nested records are mapped to the same type.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.4 (https://github.com/jOOQ/jOOQ/issues/12458)

**Comment by lukaseder** [comment]
Thanks again for your help and the MCVE!
